### ISSUE TITLE
XmlDocument and XDocument .Load badly resolves the path

### ISSUE THREAD

**Issue by jakubsuchybio** [dotnet/corefx#25820]
I am porting our server console app from .NET Framework 4.6.1 to .NET Core and after I made it compile, then I got a lot of runtime exceptions of XmlDocument and XDocument not loading.

I made a quick repro with this:
![https://i.imgur.com/79hOFVR.png](https://i.imgur.com/79hOFVR.png)
```c#
    class Program
    {
        static void Main(string[] args)
        {
            var doc = new XmlDocument();
            doc.Load("test.xml");
            var doc2 = XDocument.Load("test.xml");
        }
    }
```

The problem is, when this file `test.xml` is located at path that contains character `#`. 
For me it is at `"C:\\dev\\c#\\Test\\ConsoleApp1"` and that produces the above error.

**Comment by danmosemsft** [comment]
I'm not sure where things go wrong but XmlTextReader constructs uts URI over the string `file://c:\dev\c#\test\test.xml` which seems fine but ends up with 
```
		BaseURI	"file:///c:/dev/c#\\test\\test.xml"	string
```
internally on ParsingState is the actual Uri object
```
+		reader._impl._ps	{System.Xml.XmlTextReaderImpl.ParsingState}	System.Xml.XmlTextReaderImpl.ParsingState

-		baseUri	{file:///c:/dev/c#\test\test.xml}	System.Uri
		AbsoluteUri	"file:///c:/dev/c#%5Ctest%5Ctest.xml"	string
```
I am not sure why there's a mixture of forward and backward slashes. 

@rmkerr could this be another URI issue like https://github.com/dotnet/corefx/issues/25519? 

I didn't check what it does on desktop.

@krwq

**Comment by rmkerr** [comment]
The missing section of the path is definitely being cut out by System.Uri, but I'm not sure yet whether or not that behavior is incorrect. It looks to me like the unencoded '#' is interpreted as the beginning of a [fragment](https://tools.ietf.org/html/rfc3986#section-3.5). That means that the rest of the URI is (correctly) dropped from LocalPath. When constructing a URI that contains special characters like '#' we'd normally expect them to be percent encoded. So if we instead construct a URI with the string:

> @"file://c:\dev\c%23\test\test.xml"

We get the expected local path:

> "c:\\dev\\c#\\test\\test.xml"

**Comment by krwq** [comment]
@rmkerr is this a convergence between netfx? I'm seeing on desktop:
```csharp
Console.WriteLine(new Uri(@"c:\dev\c#\test\test.xml"))
```

produces `file:///c:/dev/c%23/test/test.xml` which seems as an expected Uri.

Xml currently parses Uri argument as seen here:
https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L36

(if uri is relative it calls `new Uri(Path.GetFullPath(relativeUri))`).

As I understand `#` inside the Uri according to spec should be dropped which means we have to now choose compat vs spec (both in XML and Uri), is that correct?

**Comment by rmkerr** [comment]
It'd be more accurate to look at LocalPath than at ToString. So for the following code:

`Uri uri1 = new Uri(@"c:\dev\c#\test\test.xml");`
`Console.WriteLine(uri1.LocalPath);`

We get a consistent result between netfx and corefx:
> "c:\dev\c#\test\test.xml"

We have a little bit of flexibility in how we treat [LocalPath](https://docs.microsoft.com/en-us/dotnet/api/system.uri.localpath?view=netcore-2.0#System_Uri_LocalPath), since it's different than the other properties of System.Uri. It produces a "local operating-system representation of a file name", which means the output might not actually be a valid URI. So it might be possible to deal with on the URI side.

A change like that would require a lot more research though, as it definitely has the potential to break things.

**Comment by karelz** [comment]
This seems to be fixed in 2.1 by PR #21826

.NET Core 2.0 is broken - see [release/2.0.0 branch code](https://github.com/dotnet/corefx/blob/release/2.0.0/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L43):
https://github.com/dotnet/corefx/blob/63e70fa6b1bc6b7175467e0ad16cbbf4c44f8d83/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L43

[master (2.1)](https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L43) is fixed:
https://github.com/dotnet/corefx/blob/f2df042b1b1772920286e0d3a87830315d6c2562/src/System.Private.Xml/src/System/Xml/XmlResolver.cs#L43

Desktop (reference source) matches what is in master:
http://referencesource.microsoft.com/#System.Xml/System/Xml/XmlResolver.cs,53

Maybe we should add a regression test case for XML with `#` in path?
